File: 프로그래머스/Level3/네트워크.java
import java.util.*;

class Solution {
    
    static ArrayList<Integer>[] graph;
    static boolean[] visited;
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        graph = new ArrayList[n+1];
        for(int i=1; i<=n; i++){
            graph[i] = new ArrayList<>();
        }
        
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(i==j)
                    continue;
                if(computers[i][j]==1){
                    graph[i+1].add(j+1);
                    graph[j+1].add(i+1);
                }
            }
        }
        
        visited = new boolean[n+1];
        for(int i=1; i<=n; i++){
            if(!visited[i])
                answer+=bfs(i,n);
        }
        return answer;
    }
    
    private int bfs(int start,int n){
        visited[start]=true;
        
        Queue<Integer> q = new LinkedList<>();
        q.offer(start);
        
        int cnt=1;
        while(!q.isEmpty()){
            int cur = q.poll();
            for(int next : graph[cur]){
                if(!visited[next])
                {
                    visited[next]=true;
                    q.offer(next);
                }
            }
        }
        return cnt;
    }
}

File: Kotlin/프로그래머스/네트워크.kt
import java.util.*;

class Solution {

    lateinit var visited: Array<Boolean>

    fun solution(n: Int, computers: Array<IntArray>): Int {
        var cnt:Int = 0
        var q = LinkedList<Int>();

        visited = Array(n,{false})
        for (i in 0 until n){
            if(!visited[i]){
                cnt++
                q.offer(i)
                visited[i] = true

                while (!q.isEmpty()){
                    var now: Int = q.poll()
                    for (j in 0 until n){
                        if(j!=now && !visited[j] && computers[now][j]==1)
                        {
                            visited[j] = true
                            q.offer(j)
                        }
                    }
                }
            }
        }

        return cnt
    }
}

fun main(args: Array<String>){
    println(Solution().solution(3, arrayOf(intArrayOf(1,1,0), intArrayOf(1,1,0), intArrayOf(0,0,1))))
}